### PR TITLE
Release v0.15 - RC0

### PR DESCRIPTION
**Pull request by anxolin**
Release, that includes Parawap changes.

Release 0.15 was not ready to deploy, and these changes are important. to not delay the version we are one exception on our release process, and we are doing a release starting from master (to avoid bringing some bugs we have in dev, and some unfinished functionalities)

This Branch therefore is MASTER, with these PRs commits cherry-picked: #791 #798

### CODE REVIEW

**Comment by anxolin** [comment]
Merging for doing a hotfix in master.

Pls, still review.

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 1]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
 
 type QuoteResult = [PromiseSettledResult<PriceInformation>, PromiseSettledResult<FeeInformation>]
 
+// const FEE_EXCEEDS_FROM_ERROR = new QuoteError({
+//   errorType: QuoteErrorCodes.FeeExceedsFrom,
+//   description: QuoteError.quoteErrorDetails.FeeExceedsFrom
+// })
+
+class PriceQuoteError extends Error {
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:46]: why is this here and not in errors?

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 2]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+    PRICE_API_TIMEOUT_MS,
+    'Paraswap: Get Price API'
+  )
+
+  // Get results from API queries
+  const [priceResult, paraSwapPriceResult] = await Promise.allSettled([pricePromise, paraSwapPricePromise])
+
+  // Prepare an array with all successful estimations
+  const priceQuotes: Array<PriceInformationWithSource> = []
+  const errorsGetPrice: Array<PromiseRejectedResultWithSource> = []
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:79]: why not `priceQuotesErrors` or is it different?

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 3]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+type PriceInformationWithSource = PriceInformation & { source: PriceSource; data?: any }
+type PromiseRejectedResultWithSource = PromiseRejectedResult & { source: PriceSource }
+
+/**
+ *
+ * @returns The best price among GPv2 API or Paraswap one
+ */
+async function _getBestPriceQuote(params: PriceQuoteParams): Promise<PriceInformation> {
+  // Get price from all API: Gpv2 and Paraswap
+  const pricePromise = withTimeout(getPriceQuote(params), PRICE_API_TIMEOUT_MS, 'GPv2: Get Price API')
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:67]: nit: i would name this `const gpPricePromise` or `gnosisProtocolPricePromise` long but explicit

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 4]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+
+  // Get results from API queries
+  const [priceResult, paraSwapPriceResult] = await Promise.allSettled([pricePromise, paraSwapPricePromise])
+
+  // Prepare an array with all successful estimations
+  const priceQuotes: Array<PriceInformationWithSource> = []
+  const errorsGetPrice: Array<PromiseRejectedResultWithSource> = []
+
+  if (isPromiseFulfilled(priceResult)) {
+    priceQuotes.push({ ...priceResult.value, source: 'gnosis-protocol' })
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:82]: nit: the rename i suggested above makes this a breeze to read, ah it's the gnosis protocol price quotes being pushed, took me a seconds

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 5]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+
+  if (isPromiseFulfilled(paraSwapPriceResult)) {
+    const paraswapPrice = toPriceInformation(paraSwapPriceResult.value)
+    paraswapPrice && priceQuotes.push({ ...paraswapPrice, source: 'paraswap', data: paraSwapPriceResult.value })
+  } else {
+    errorsGetPrice.push({ ...paraSwapPriceResult, source: 'paraswap' })
+  }
+
+  if (errorsGetPrice.length > 0) {
+    const sourceNames = errorsGetPrice.map(e => e.source).join(', ')
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:95]: nit: this map feels unnecessary for logging, i would just print the array and we deal with it ourselves when debugging in chrome tools

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 6]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+    console.error('[hooks::useRefetchPriceCallback] Some API failed or timed out: ' + sourceNames, errorsGetPrice)
+  }
+
+  if (priceQuotes.length > 0) {
+    const sourceNames = priceQuotes.map(p => p.source).join(', ')
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:100]: is there such thing a successful price query if there is no amount? Asking about the logic in L102 and related to this. Why not use a reduce to loop/filter the data once and get sourceNames at the same time. If the amount is empty as you're checking in L102 you don't want the source name correct?

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 7]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+    console.log('[hooks::useRefetchPriceCallback] Get best price succeeded for ' + sourceNames, priceQuotes)
+    const amounts = priceQuotes.map(quote => quote.amount).filter(Boolean) as string[]
+
+    // Take the best price: Aggregate all the amounts into a single one.
+    //  - Use maximum of all the result for "Sell orders":
+    //        You want to get the maximum number of buy tokens
+    //  - Use minimum "Buy orders":
+    //        You want to spend the min number of sell tokens
+    const aggregationFunction = params.kind === OrderKind.SELL ? 'max' : 'min'
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:109]: nit: this isn't a function, i would name this `aggregationType` or `aggregationMethodName`

File: src/custom/hooks/useRefetchPriceCallback.ts [hunk 8]
```diff
@@ -27,6 +38,91 @@ export interface RefetchQuoteCallbackParmams {
...
+    const amount = BigNumberJs[aggregationFunction](...amounts).toString(10)
+    const token = priceQuotes[0].token
+    // console.log('Aggregated amounts', aggregationFunction, amounts, amount)
```
Comment (W3stside) [inline src/custom/hooks/useRefetchPriceCallback.ts:112]: delete?

File: src/custom/utils/paraswap.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+): rateResult is OptimalRatesWithPartnerFees {
+  return !!(rateResult as OptimalRatesWithPartnerFees).destAmount
+}
+
+export async function getPriceQuote(params: PriceQuoteParams): Promise<ParaSwapPriceQuote | null> {
+  const { baseToken: baseTokenAux, quoteToken: quoteTokenAux, amount, kind, chainId } = params
+  const baseToken = toErc20Address(baseTokenAux, chainId)
+  const quoteToken = toErc20Address(quoteTokenAux, chainId)
+
+  let paraSwap = paraSwapLibs.get(chainId)
```
Comment (W3stside) [inline src/custom/utils/paraswap.ts:61]: nit: i would move this logic above the `const` declaration for base/quoteToken in case it returns null. you don't use the constants until later anyways
